src/match_return.jl

```diff
@@ -76,24 +76,25 @@ end
 # in which `MatchFailure` is a possible result.
 #
 function adjust_case_for_return_macro(__module__, location, pattern, result, predeclared_temps)
-    value = gensym("value")
-    label = gensym("label")
     found_early_exit::Bool = false
+
+    # Check for the presence of early exit macros @match_return and @match_fail
     function adjust_top(p)
         is_expr(p, :macrocall) || return p
         if length(p.args) == 3 &&
             (p.args[1] == :var"@match_return"  || p.args[1] == Expr(:., Symbol(string(@__MODULE__)), QuoteNode(:var"@match_return")))
             # :(@match_return e) -> :($value = $e; @goto $label)
             found_early_exit = true
-            return Expr(:block, p.args[2], :($value = $(p.args[3])), :(@goto $label))
+            # expansion of the result will be done later by ExpressionRequiringAdjustmentForReturnMacro 
+            return p
         elseif length(p.args) == 2 &&
             (p.args[1] == :var"@match_fail"  || p.args[1] == Expr(:., Symbol(string(@__MODULE__)), QuoteNode(:var"@match_fail")))
             # :(@match_fail) -> :($value = $MatchFaulure; @goto $label)
             found_early_exit = true
-            return Expr(:block, p.args[2], :($value = $MatchFailure), :(@goto $label))
+            # expansion of the result will be done later by ExpressionRequiringAdjustmentForReturnMacro 
+            return p
         elseif length(p.args) == 4 &&
-            (p.args[1] == :var"@match" || p.args[1] == Expr(:., Symbol(string(@__MODULE__)), QuoteNode(:var"@match")) ||
-             p.args[1] == :var"@match" || p.args[1] == Expr(:., Symbol(string(@__MODULE__)), QuoteNode(:var"@match")))
+            (p.args[1] == :var"@match" || p.args[1] == Expr(:., Symbol(string(@__MODULE__)), QuoteNode(:var"@match")))
```

Comment:
Not a big deal, but can you factor this pattern out into a function?

Reply:
Yes